libraries/chain_kv/include/b1/session/undo_stack.hpp

```diff
@@ -197,4 +215,123 @@ typename undo_stack<Session>::const_variant_type undo_stack<Session>::bottom() c
+         uint32_t version = 0;
+         fc::raw::unpack( ds, version );
+         EOS_ASSERT( version >= undo_stack_min_supported_version && version <= undo_stack_max_supported_version,
+                    eosio::chain::chain_exception,
+                    "Unsupported version of Undo stack data file '${filename}'. "
+                    "Undo stack data version is ${version} while code supports version(s) [${min},${max}]",
+                    ("filename", undo_stack_dat.generic_string())
+                    ("version", version)
+                    ("min", undo_stack_min_supported_version)
+                    ("max", undo_stack_max_supported_version)
+         );
+
+         int64_t rev; fc::raw::unpack( ds, rev );
+         revision(rev);
+
+         size_t num_sessions; fc::raw::unpack( ds, num_sessions );
+         for( size_t i = 0; i < num_sessions; ++i ) {
+            session_type session;
+            if (m_sessions.empty())
+               session.attach(*m_head);
+            else
+               session.attach(m_sessions.back());
+
+            size_t num_updated_keys; fc::raw::unpack( ds, num_updated_keys );
+            for( size_t j = 0; j < num_updated_keys; ++j ) {
+               shared_bytes_reflect raw_key; fc::raw::unpack( ds, raw_key );
+               shared_bytes_reflect raw_value; fc::raw::unpack( ds, raw_value );
+
+               shared_bytes key(raw_key.data.data(), raw_key.data.size());
+               shared_bytes value(raw_value.data.data(), raw_value.data.size());
+               session.write(key, value);
+            }
+
+            size_t num_deleted_keys; fc::raw::unpack( ds, num_deleted_keys );
+            for( size_t j = 0; j < num_deleted_keys; ++j ) {
+               shared_bytes_reflect raw_key; fc::raw::unpack( ds, raw_key );
+
+               shared_bytes key(raw_key.data.data(), raw_key.data.size());
+               session.erase(key);
+            }
+
+            m_sessions.emplace_back(std::move(session));
+         }
+      } FC_CAPTURE_AND_RETHROW( (undo_stack_dat) )
+
+      fc::remove( undo_stack_dat );
+   }
+}
+
+template <typename Session>
+void undo_stack<Session>::close() {
+   if (m_datadir.empty())
+      return;
+
+   auto undo_stack_dat = m_datadir / undo_stack_filename;
+
+   std::ofstream out( undo_stack_dat.generic_string().c_str(), std::ios::out | std::ios::binary | std::ofstream::trunc );
+   fc::raw::pack( out, undo_stack_magic_number );
+   fc::raw::pack( out, undo_stack_max_supported_version );
+
+   fc::raw::pack( out, revision() );
+   fc::raw::pack( out, size() ); // number of sessions
+
+   while ( !m_sessions.empty() ) {
+      auto& session = m_sessions.front();
+      auto updated_keys = session.updated_keys();
+      fc::raw::pack( out, updated_keys.size() ); // number of updated keys
+
+      for (const auto& key: updated_keys) {
+        auto value = session.read(key);
+
+        if ( value ) {
+           shared_bytes_reflect raw_key; 
```

Comment:
with the datastream operators this would become...
```c++
out << key;
out << value;
```

Reply:
Changed.